api/workload/v3/workload.go

```diff
@@ -0,0 +1,218 @@
+func Addr(addr string) Option {
+	return func(w *X509SVIDClient) {
+		w.addr = addr
+	}
+}
+
+// Logger specifies the logger to use.
+func Logger(logger *logrus.Logger) Option {
+	return func(w *X509SVIDClient) {
+		w.logger = logger
+	}
+}
+
+// NewX509SVIDClient returns a new Workload API client for X.509 SVIDs.
+func NewX509SVIDClient(watcher WorkloadX509SVIDWatcher, opts ...Option) (*X509SVIDClient, error) {
+	ctx, cancel := context.WithCancel(context.Background())
+	c := &X509SVIDClient{
+		logger:   logrus.StandardLogger(),
+		addr:     DefaultAgentAddress,
+		watcher:  watcher,
+		ctx:      ctx,
+		cancelFn: cancel,
+		backoff:  newBackoff(),
+	}
+	for _, opt := range opts {
+		opt(c)
+	}
+	if !strings.HasPrefix(c.addr, _unixPathPrefix) {
+		return nil, fmt.Errorf("spiffe/workload: agent address %q is not a unix address", c.addr)
+	}
+	return c, nil
+}
+
+// Start starts the client.
+//
+// The client will always start, and users should rely on the watcher
+// interface to receives updates on the client's status.
+func (c *X509SVIDClient) Start(ctx context.Context) error {
+	c.wg.Add(1)
+	go c.run()
+	return nil
+}
+
+func (c *X509SVIDClient) run() {
+	defer c.wg.Done()
+	for {
+		if done := c.watch(); done {
+			return
+		}
+	}
+}
+
+// creates single watch loop and returns whether we should continue watching
+func (c *X509SVIDClient) watch() (done bool) {
+	conn, err := c.newConn()
+	if err != nil {
+		return c.handleError(err)
+	}
+	c.watcher.OnConnection(true)
+	defer c.watcher.OnConnection(false)
+	defer conn.Close()
+
+	stream, err := c.newX509SVIDStream(conn)
+	if err != nil {
+		return c.handleError(err)
+	}
+	if err := c.handleX509SVIDStream(stream); err != nil {
+		return c.handleError(err)
+	}
+	return false
+}
+
+func (c *X509SVIDClient) newConn() (*grpc.ClientConn, error) {
+	conn, err := grpc.DialContext(c.ctx, c.addr, grpc.WithInsecure(), grpc.WithBlock())
+	if err != nil {
+		return nil, fmt.Errorf("spiffe/workload: failed to dial workload API at %q: %v", c.addr, err)
+	}
+	return conn, nil
+}
+
+// handles an error, applies backoff, and returns true if the context has been canceled
+func (c *X509SVIDClient) handleError(err error) (done bool) {
+	if status.Code(err) == codes.Canceled {
+		return true
+	}
+	c.watcher.OnError(err)
+	select {
+	case <-time.After(c.backoff.Duration()):
+		return false
+	case <-c.ctx.Done():
+		return true
+	}
+}
+
+func (c *X509SVIDClient) newX509SVIDStream(conn *grpc.ClientConn) (workload.SpiffeWorkloadAPI_FetchX509SVIDClient, error) {
+	workloadClient := workload.NewSpiffeWorkloadAPIClient(conn)
+	header := metadata.Pairs("workload.spiffe.io", "true")
+	grpcCtx := metadata.NewOutgoingContext(c.ctx, header)
+	return workloadClient.FetchX509SVID(grpcCtx, &workload.X509SVIDRequest{})
+}
+
+func (c *X509SVIDClient) handleX509SVIDStream(stream workload.SpiffeWorkloadAPI_FetchX509SVIDClient) error {
+	for {
+		resp, err := stream.Recv()
```

Comment:
Am I correct in saying that for this to terminate we depend on grpc.WithBlock to enable the context to control the conn which will cancel the stream?

I guess the grpcCtx above is coming from that c.ctx so maybe we are okay either way.

Reply:
yeah this gets canceled because of the grpcCtx that chains up to the main context